### DISCUSSION TITLE
Serializing nested objects

### DISCUSSION THREAD

**Discussion by obsgolem** [wasm-bindgen/wasm-bindgen#4194] [Q&A]
### Summary

Is there a canonical way to serialize a struct which contains another `JsValue`?

### Additional Details
I am trying to implement bindings for https://cesium.com/learn/cesiumjs/ref-doc/Viewer.html, and am running into issues creating bindings for the `ConstructorOptions`  object. This object you need to be able to provide an `ImageryProvider`, which itself is a javascript object which needs to be generated by a constructor. However, since `JsValue` does not implement `Serialize` you can't `#[derive(Serialize)]` for a struct which contains an `ImageryProvider`. My question is what is the canonical way to handle transporting such data from Rust to Javascript?

**Comment by obsgolem** [comment]
I hacked around this using the following macro which I modified from the `hashlit` crate:
```rust
macro_rules! object {
    ($($key:expr => $value:expr,)+) => { object!($($key => $value),+) };
    ($($key:expr => $value:expr),*) => {
        {
            let mut _obj = ::js_sys::Object::new();
            $(
                js_sys::Reflect::set(&_obj, &($key.into()), &($value.into())).unwrap();
            )*
            _obj
        }
    };
}
```
If there is no better way to do this kind of thing, then this might be useful to build into the crate.

**Comment by DennisJensen95** [comment]
Bump, I would very much like an answer to the same 👍

**Comment by daxpedda** [comment]
You might want to take a look at [`serde-wasm-bindgen`](https://crates.io/crates/serde-wasm-bindgen).

Alternatively, the most straightforward way would be to use [`JSON.stringify()` and `JSON.parse()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON), which is also what is used by the deprecated Serde feature of `wasm-bindgen`.
